src/multi_vector_simulator/E4_verification.py

```diff
@@ -154,3 +156,50 @@ def detect_excessive_excess_generation_in_bus(dict_values):
             if ratio < 0.9:
                 msg = f"Attention, on bus {bus_label} there is excessive excess generation, totalling up to {round((1 - ratio) * 100)}% of the inflows. The total inflows are {round(total_inflow_bus)} and outflows {round(total_outflow_bus)}  It seems to be cheaper to have this excess generation than to install more capacities that forward the energy carrier to other busses (if those assets can be optimized)."
                 logging.warning(msg)
+
+
+def verify_state_of_charge(dict_values):
+    r"""
+    This function checks the state of charge of each storage component
+    It raises warning log messages if the SoC has a physically infeasible value
+
+    Parameters
+    ----------
+    dict_values
+
+    Returns
+    -------
+    - Nothing if there are no physically infeasible SoC values for the storage components
+    - Prints log messages to console and log file if there are physically impossible SoC values
+
+    Notes
+    -----
+    Tested with:
+    - test_E4_verification.test_verify_state_of_charge_feasible()
+    - test_E4_verification.test_verify_state_of_charge_soc_below_zero()
+    - test_E4_verification.test_verify_state_of_charge_soc_above_zero()
+
+    """
+    # Dict holding the data of all of the storage components
+    storage_data_dict = dict_values[ENERGY_STORAGE]
+
+    # Check if storage components are present in the energy system and only then verify SoC values
+    if bool(storage_data_dict) is not False:
+        # Loop through the storage components
+        for storage in list(storage_data_dict.keys()):
+            # Get the SoC time-series in a pandas series data structure
+            soc_timeseries = storage_data_dict[storage][TIMESERIES_SOC]
+            # Loop through the time-steps of the SoC time series
+            for index, soc_value in soc_timeseries.items():
```

Comment:
This is a loop though each single entry of the SOC timeseries. That can make it more work intensive for your PC, but also potentially results in a lot of error messages. We can let it pass, as we dont expect that the test will ever show us this unintended outcome (SOC < 0, SOC > 1), but I think it would be cleaner to do:

```
if (soc_timeseries > 1).any():
    logging.warning(
        f"SoC of {storage} has at least one time step where its values is  greater than 1.   This is a physically impossible value!"
        )
```

Reply:
Aha, you wanted only one warning irrespective of the number of times it happens, correct?

Reply:
But since I wrote tests, keeping my method in mind.. can we leave it like this? :/

Reply:
But this would not change the tests? I mean, sure, you need to change the message that is asserted (you can also check for part of the message), but only that.